bpf/tests/tc_egressgw_snat.c

```diff
@@ -149,87 +150,185 @@ int egressgw_snat_check(const struct __ctx_buff *ctx)
-		test_fatal("dst IP has changed");
+		if (l3->daddr != CLIENT_IP)
+			test_fatal("dst IP hasn't been revSNATed to client IP");
+	} else {
+		if (memcmp(l2->h_source, (__u8 *)client_mac, ETH_ALEN) != 0)
+			test_fatal("src MAC is not the client MAC")
+
+		if (memcmp(l2->h_dest, (__u8 *)ext_svc_mac, ETH_ALEN) != 0)
+			test_fatal("dst MAC is not the external svc MAC")
+
+		if (l3->saddr != EGRESS_IP)
+			test_fatal("src IP hasn't been NATed to egress gateway IP");
+
+		if (l3->daddr != EXTERNAL_SVC_IP)
+			test_fatal("dst IP has changed");
+	}
 
 	/* Lookup the SNAT mapping for the original packet to determine the new source port */
 	struct ipv4_ct_tuple tuple = {
-		.daddr   = EXTERNAL_SVC_IP,
-		.saddr   = CLIENT_IP,
+		.daddr   = CLIENT_IP,
+		.saddr   = EXTERNAL_SVC_IP,
 		.dport   = EXTERNAL_SVC_PORT,
 		.sport   = CLIENT_PORT,
 		.nexthdr = IPPROTO_TCP,
+		.flags = TUPLE_F_OUT,
 	};
+	struct ct_entry *ct_entry = map_lookup_elem(get_ct_map4(&tuple), &tuple);
+
+	if (!ct_entry)
+		test_fatal("no CT entry found");
+
+	if (ct_entry->tx_packets != tx_packets)
+		test_fatal("bad TX packet count (expected %u, actual %u)",
+			   tx_packets, ct_entry->tx_packets)
+	if (ct_entry->rx_packets != rx_packets)
+#ifdef ISSUE_25110_FIXED
+		/* This test fails until this issue is fixed:
+		 * https://github.com/cilium/cilium/issues/25110
+		 */
+		test_fatal("bad RX packet count (expected %u, actual %u)",
+			   rx_packets, ct_entry->rx_packets)
+#else
+		;
+#endif
+
+	tuple.saddr = CLIENT_IP;
+	tuple.daddr = EXTERNAL_SVC_IP;
 
 	struct ipv4_nat_entry *nat_entry = __snat_lookup(&SNAT_MAPPING_IPV4, &tuple);
 
 	if (!nat_entry)
 		test_fatal("could not find a NAT entry for the packet");
 
-	if (l4->source != nat_entry->to_sport)
-		test_fatal("src TCP port hasn't been NATed to egress gateway port");
+	if (reply) {
+		if (l4->source != EXTERNAL_SVC_PORT)
+			test_fatal("src port has changed");
 
-	if (l4->dest != EXTERNAL_SVC_PORT)
-		test_fatal("dst port has changed");
+		if (l4->dest != CLIENT_PORT)
+			test_fatal("dst TCP port hasn't been revSNATed to client port");
+	} else {
+		if (l4->source != nat_entry->to_sport)
+			test_fatal("src TCP port hasn't been NATed to egress gateway port");
+
+		if (l4->dest != EXTERNAL_SVC_PORT)
+			test_fatal("dst port has changed");
+	}
 
 	test_finish();
 }
 
-/* Test that a packet matching an excluded CIDR egress gateway policy on the
- * to-netdev program does not get SNATed with the egress IP of the policy.
+/* Test that a packet matching an egress gateway policy on the to-netdev program
+ * gets correctly SNATed with the egress IP of the policy.
  */
-PKTGEN("tc", "tc_egressgw_skip_excluded_cidr_snat")
-int egressgw_skip_excluded_cidr_snat_pktgen(struct __ctx_buff *ctx)
+PKTGEN("tc", "tc_egressgw_snat1")
+int egressgw_snat1_pktgen(struct __ctx_buff *ctx)
 {
-	struct pktgen builder;
-	struct tcphdr *l4;
-	struct ethhdr *l2;
-	struct iphdr *l3;
-	void *data;
+	return egressgw_snat_pktgen(ctx, false);
```

Comment:
nit: having an enum (or just defining a couple of constants) would help figuring out more quickly what we are testing